File: engine/src/main/java/de/hanno/hpengine/engine/graphics/renderer/drawstrategy/extensions/PixelPerfectPickingExtension.java
package de.hanno.hpengine.engine.graphics.renderer.drawstrategy.extensions;

import de.hanno.hpengine.engine.backend.Backend;
import de.hanno.hpengine.engine.backend.OpenGl;
import de.hanno.hpengine.engine.event.MeshSelectedEvent;
import de.hanno.hpengine.engine.graphics.GpuContext;
import de.hanno.hpengine.engine.graphics.state.RenderState;
import de.hanno.hpengine.engine.graphics.renderer.drawstrategy.FirstPassResult;
import org.lwjgl.BufferUtils;
import org.lwjgl.opengl.GL11;
import org.joml.Vector2f;

import java.nio.FloatBuffer;
import java.util.logging.Logger;

public class PixelPerfectPickingExtension implements RenderExtension<OpenGl> {

    private final FloatBuffer floatBuffer;

    public PixelPerfectPickingExtension() {
        // 4 channels
        floatBuffer = BufferUtils.createFloatBuffer(4);
    }
    @Override
    public void renderFirstPass(Backend<OpenGl> backend, GpuContext<OpenGl> gpuContext, FirstPassResult firstPassResult, RenderState renderState) {

        if (backend.getInput().getPickingClick() == 1) {
            gpuContext.readBuffer(4);

            floatBuffer.rewind();
//             TODO: This doesn't make sense anymore, does it?
            Vector2f ratio = new Vector2f((float) gpuContext.getWindow().getWidth() / (float) gpuContext.getWindow().getWidth(),
                    (float) gpuContext.getWindow().getHeight() / (float) gpuContext.getWindow().getHeight());
            int adjustedX = (int) (backend.getInput().getMouseX() * ratio.x);
            int adjustedY = (int) (backend.getInput().getMouseY() * ratio.y);
            GL11.glReadPixels(adjustedX, adjustedY, 1, 1, GL11.GL_RGBA, GL11.GL_FLOAT, floatBuffer);
            Logger.getGlobal().info("Picked: " + adjustedX + " : " + adjustedY);
            try {
                int entityIndexComponentIndex = 0; // red component
                int meshIndexComponentIndex = 3; // alpha component
                int entityIndex = (int) floatBuffer.get(entityIndexComponentIndex);
                int meshIndex = (int) floatBuffer.get(meshIndexComponentIndex);
                backend.getEventBus().post(new MeshSelectedEvent(entityIndex, meshIndex));
            } catch (Exception e) {
                e.printStackTrace();
            }
            backend.getInput().setPickingClick(0);
        }
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/graphics/renderer/drawstrategy/extensions/PixelPerfectPickingExtension.kt
package de.hanno.hpengine.graphics.renderer.drawstrategy.extensions

import de.hanno.hpengine.config.Config
import de.hanno.hpengine.graphics.GraphicsApi
import de.hanno.hpengine.graphics.renderer.constants.Format
import de.hanno.hpengine.graphics.state.RenderState
import de.hanno.hpengine.input.Input
import de.hanno.hpengine.input.MouseClickListener
import org.joml.Vector2f
import org.joml.Vector2i
import org.lwjgl.BufferUtils
import java.nio.FloatBuffer

context(GraphicsApi)
class PixelPerfectPickingExtension(
    private val config: Config,
    private val input: Input,
    private val listeners: List<OnClickListener>
) : DeferredRenderExtension {
    override val renderPriority: Int = 1000
    private val floatBuffer: FloatBuffer = BufferUtils.createFloatBuffer(4)

    private val mouseClickListener = MouseClickListener(input)

    override fun update(deltaSeconds: Float) {
        mouseClickListener.update(deltaSeconds)
    }
    override fun renderFirstPass(
        renderState: RenderState
    ) {
        mouseClickListener.consumeClick {
            readBuffer(4)
            floatBuffer.rewind()
            val ratio = Vector2f(
                config.width.toFloat() / window.width.toFloat(),
                config.height.toFloat() / window.height.toFloat()
            )
            val adjustedX = (input.getMouseX() * ratio.x).toInt()
            val adjustedY = (input.getMouseY() * ratio.y).toInt()
            readPixels(adjustedX, adjustedY, 1, 1, Format.RGBA, floatBuffer)
            try {
                val entityId = floatBuffer[0].toInt()
                val entityBufferIndex = floatBuffer[1].toInt()
                val materialIndex = floatBuffer[2].toInt()
                val meshIndex = floatBuffer[3].toInt()
                val indices = Indices(entityBufferIndex, entityId, meshIndex, materialIndex)
                listeners.forEach {
                    it.onClick(Vector2i(adjustedX, adjustedY), indices)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}